### DISCUSSION TITLE
Migrating from UWP to Windows App SDK, how to configure ui thread for StoreContext

### DISCUSSION THREAD

**Discussion by lospeq** [microsoft/WindowsAppSDK#1866] [Q&A] [answered]
I'm trying to migrate a web app from uwp, it uses webview (now webview2 in winui3) and has some IAPs done through Windows.Services.Store.StoreContext. 

[Documentation](https://docs.microsoft.com/en-us/windows/uwp/monetize/in-app-purchases-and-trials#desktop) has a section about configuring StoreContext for the use with bridge and handling the app's main window, but all calls to the store sdk from the UWP app were working just fine without this configuration. 

However, the same calls from WinUI3 app throw a `0x80070578 (ERROR_INVALID_WINDOW_HANDLE)` [exception](https://docs.microsoft.com/en-us/uwp/api/windows.services.store.storecontext.requestpurchaseasync?view=winrt-22000), which I guess is natural given that WinUI3 runs from a MainWindow in contrast to UWP's MainPage.

But when I add StoreContext configuration code from docs above it crashes on 
`IInitializeWithWindow initWindow = (IInitializeWithWindow)(object)context;`
 part of the config with `Unable to cast from 'windows services.store.storecontext' to type 'IInitializeWithWindow' `exeption.
 
 How do I go around these issues and use StoreContext correctly, if it is possible at all?

**Comment by andrewleader** [comment]
Thanks for reporting this @lospeq, we're looking into this right now!

**Comment by andrewleader** [comment, marked as answer]
Hi @lospeq, the docs for InitializeWithWindow are a bit outdated, the code in the docs works for .NET Core 3 apps but not .NET 5/6 apps.

**Here's what you should use instead!** Use the `WinRT.Interop.InitializeWithWindow.Initialize(...)` method as seen below... no need for defining the interface with the GUID anymore 😊

```csharp
// Get the StoreContext
var storeContext = Windows.Services.Store.StoreContext.GetDefault();

// Get your HWND
var hwnd = WinRT.Interop.WindowNative.GetWindowHandle(m_window);

// And set the HWND on the StoreContext
WinRT.Interop.InitializeWithWindow.Initialize(storeContext, hwnd);

// Then use like normal!
```

We'll get the docs updated!

**Reply by lospeq** [reply]
Thank you, also found out that simply changing the type cast line from old docs to 
```
using WinRT;
var initWindow = context.As<IInitializeWithWindow>();
```
seems to also do the trick, but dropping interface definition altogether sure looks cleaner.
